File: src/browser.js
/**
 * Browser module foundations for Phase 2.
 */

export const SLATE_MAKER = 'VOIDENMARK';

export class BrowserTab {
  /**
   * @param {{ id: string, url: string, title?: string, isPinned?: boolean, isActive?: boolean }} input
   */
  constructor(input) {
    if (!input?.id || !input?.url) {
      throw new Error('BrowserTab requires both id and url.');
    }

    this.id = input.id;
    this.url = input.url;
    this.title = input.title ?? input.url;
    this.isPinned = Boolean(input.isPinned);
    this.isActive = Boolean(input.isActive);
  }

  toJSON() {
    return {
      id: this.id,
      url: this.url,
      title: this.title,
      isPinned: this.isPinned,
      isActive: this.isActive
    };
  }
}

export class BrowserModule {
  constructor() {
    /** @type {Map<string, BrowserTab>} */
    this.tabs = new Map();
    /** @type {{ url: string, title: string, visitedAt: string }[]} */
    this.history = [];
    /** @type {Map<string, { url: string, title: string, createdAt: string }>} */
    this.bookmarks = new Map();
    /** @type {{ id: string, url: string, status: 'queued' | 'in_progress' | 'completed' | 'failed', savedPath: string | null }[]} */
    this.downloads = [];
    /** @type {Set<string>} */
    this.adBlockRules = new Set();
  }

  /** @param {{ id: string, url: string, title?: string, isPinned?: boolean }} input */
  openTab(input) {
    if (this.tabs.has(input.id)) {
      throw new Error(`Tab with id "${input.id}" already exists.`);
    }

    const nextTab = new BrowserTab({ ...input, isActive: false });
    if (this.tabs.size === 0) {
      nextTab.isActive = true;
    }

    this.tabs.set(nextTab.id, nextTab);
    this.recordVisit(nextTab.url, nextTab.title);
    return nextTab;
  }

  /** @param {string} tabId */
  closeTab(tabId) {
    const tab = this.tabs.get(tabId);
    if (!tab) {
      return false;
    }

    const wasActive = tab.isActive;
    this.tabs.delete(tabId);

    if (wasActive && this.tabs.size > 0) {
      const firstRemaining = this.listTabs()[0];
      firstRemaining.isActive = true;
    }

    return true;
  }

  /** @param {string} tabId */
  activateTab(tabId) {
    const target = this.tabs.get(tabId);
    if (!target) {
      throw new Error(`Unknown tab: ${tabId}`);
    }

    for (const tab of this.tabs.values()) {
      tab.isActive = false;
    }

    target.isActive = true;
    return target;
  }

  /** @param {string} tabId @param {string} url */
  navigateTab(tabId, url) {
    const tab = this.tabs.get(tabId);
    if (!tab) {
      throw new Error(`Unknown tab: ${tabId}`);
    }

    tab.url = url;
    tab.title = url;
    this.recordVisit(url, url);
    return tab;
  }

  listTabs() {
    return [...this.tabs.values()].sort((a, b) => Number(b.isPinned) - Number(a.isPinned));
  }

  /** @param {string} tabId @param {boolean} pinned */
  pinTab(tabId, pinned = true) {
    const tab = this.tabs.get(tabId);
    if (!tab) {
      throw new Error(`Unknown tab: ${tabId}`);
    }

    tab.isPinned = pinned;
    return tab;
  }

  /** @param {string} url @param {string} title */
  addBookmark(url, title = url) {
    const createdAt = new Date().toISOString();
    this.bookmarks.set(url, { url, title, createdAt });
    return this.bookmarks.get(url);
  }

  /** @param {string} url */
  removeBookmark(url) {
    return this.bookmarks.delete(url);
  }

  listBookmarks() {
    return [...this.bookmarks.values()].sort((a, b) => a.title.localeCompare(b.title));
  }

  /** @param {string} url @param {string} title */
  recordVisit(url, title = url) {
    this.history.unshift({ url, title, visitedAt: new Date().toISOString() });
    return this.history[0];
  }

  listHistory(limit = 20) {
    return this.history.slice(0, Math.max(0, limit));
  }

  /** @param {string} pattern */
  addAdBlockRule(pattern) {
    this.adBlockRules.add(pattern.toLowerCase());
    return this.adBlockRules.size;
  }

  /** @param {string} url */
  isBlocked(url) {
    const lowerUrl = url.toLowerCase();
    for (const pattern of this.adBlockRules) {
      if (lowerUrl.includes(pattern)) {
        return true;
      }
    }
    return false;
  }

  /** @param {{ id: string, url: string }} input */
  queueDownload(input) {
    if (this.downloads.some((download) => download.id === input.id)) {
      throw new Error(`Download with id "${input.id}" already exists.`);
    }

    const job = {
      id: input.id,
      url: input.url,
      status: 'queued',
      savedPath: null
    };

    this.downloads.push(job);
    return job;
  }

  /** @param {string} id @param {'queued'|'in_progress'|'completed'|'failed'} status @param {string | null} [savedPath] */
  updateDownload(id, status, savedPath = null) {
    const download = this.downloads.find((job) => job.id === id);
    if (!download) {
      throw new Error(`Unknown download: ${id}`);
    }

    download.status = status;
    download.savedPath = savedPath;
    return download;
  }

  /**
   * Lightweight webview descriptor for IPC integration.
   * @param {string} tabId
   */
  webviewConfig(tabId) {
    const tab = this.tabs.get(tabId);
    if (!tab) {
      throw new Error(`Unknown tab: ${tabId}`);
    }

    return {
      tabId: tab.id,
      src: tab.url,
      sandbox: true,
      adBlocked: this.isBlocked(tab.url)
    };
  }

  toJSON() {
    return {
      maker: SLATE_MAKER,
      tabs: this.listTabs().map((tab) => tab.toJSON()),
      bookmarks: this.listBookmarks(),
      history: [...this.history],
      downloads: [...this.downloads],
      adBlockRules: [...this.adBlockRules]
    };
  }

  /** @param {{ tabs?: Array<any>, bookmarks?: Array<any>, history?: Array<any>, downloads?: Array<any>, adBlockRules?: Array<string> }} payload */
  static fromJSON(payload = {}) {
    const browser = new BrowserModule();

    for (const tabInput of payload.tabs ?? []) {
      const tab = new BrowserTab(tabInput);
      browser.tabs.set(tab.id, tab);
    }

    for (const bookmark of payload.bookmarks ?? []) {
      browser.bookmarks.set(bookmark.url, bookmark);
    }

    browser.history = [...(payload.history ?? [])];
    browser.downloads = [...(payload.downloads ?? [])];
    browser.adBlockRules = new Set((payload.adBlockRules ?? []).map((rule) => rule.toLowerCase()));

    return browser;
  }
}

File: slate/browser.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone

SLATE_MAKER = "VOIDENMARK"


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


@dataclass(slots=True)
class BrowserTab:
    id: str
    url: str
    title: str
    is_pinned: bool = False
    is_active: bool = False


@dataclass(slots=True)
class Bookmark:
    url: str
    title: str
    created_at: str = field(default_factory=_now_iso)


@dataclass(slots=True)
class HistoryEntry:
    url: str
    title: str
    visited_at: str = field(default_factory=_now_iso)


@dataclass(slots=True)
class DownloadJob:
    id: str
    url: str
    status: str = "queued"
    saved_path: str | None = None


class BrowserModule:
    def __init__(self) -> None:
        self._tabs: dict[str, BrowserTab] = {}
        self._history: list[HistoryEntry] = []
        self._bookmarks: dict[str, Bookmark] = {}
        self._downloads: dict[str, DownloadJob] = {}
        self._adblock_rules: set[str] = set()

    @property
    def tabs(self) -> tuple[BrowserTab, ...]:
        return tuple(sorted(self._tabs.values(), key=lambda tab: (not tab.is_pinned, tab.id)))

    def open_tab(self, tab_id: str, url: str, title: str | None = None) -> BrowserTab:
        if tab_id in self._tabs:
            raise ValueError(f'Tab with id "{tab_id}" already exists.')

        tab = BrowserTab(
            id=tab_id,
            url=url,
            title=title or url,
            is_active=not self._tabs,
        )
        self._tabs[tab_id] = tab
        self.record_visit(url, tab.title)
        return tab

    def close_tab(self, tab_id: str) -> bool:
        tab = self._tabs.get(tab_id)
        if tab is None:
            return False

        was_active = tab.is_active
        del self._tabs[tab_id]

        if was_active and self._tabs:
            first = next(iter(self.tabs))
            first.is_active = True

        return True

    def activate_tab(self, tab_id: str) -> BrowserTab:
        if tab_id not in self._tabs:
            raise ValueError(f"Unknown tab: {tab_id}")

        for tab in self._tabs.values():
            tab.is_active = False

        self._tabs[tab_id].is_active = True
        return self._tabs[tab_id]

    def navigate_tab(self, tab_id: str, url: str) -> BrowserTab:
        if tab_id not in self._tabs:
            raise ValueError(f"Unknown tab: {tab_id}")

        tab = self._tabs[tab_id]
        tab.url = url
        tab.title = url
        self.record_visit(url, url)
        return tab

    def pin_tab(self, tab_id: str, pinned: bool = True) -> BrowserTab:
        if tab_id not in self._tabs:
            raise ValueError(f"Unknown tab: {tab_id}")

        self._tabs[tab_id].is_pinned = pinned
        return self._tabs[tab_id]

    def add_bookmark(self, url: str, title: str | None = None) -> Bookmark:
        bookmark = Bookmark(url=url, title=title or url)
        self._bookmarks[url] = bookmark
        return bookmark

    def remove_bookmark(self, url: str) -> bool:
        return self._bookmarks.pop(url, None) is not None

    @property
    def bookmarks(self) -> tuple[Bookmark, ...]:
        return tuple(sorted(self._bookmarks.values(), key=lambda item: item.title.lower()))

    def record_visit(self, url: str, title: str | None = None) -> HistoryEntry:
        entry = HistoryEntry(url=url, title=title or url)
        self._history.insert(0, entry)
        return entry

    def history(self, limit: int = 20) -> tuple[HistoryEntry, ...]:
        return tuple(self._history[: max(0, limit)])

    def add_adblock_rule(self, pattern: str) -> int:
        self._adblock_rules.add(pattern.lower())
        return len(self._adblock_rules)

    def remove_adblock_rule(self, pattern: str) -> bool:
        normalized = pattern.lower()
        if normalized not in self._adblock_rules:
            return False
        self._adblock_rules.remove(normalized)
        return True

    @property
    def adblock_rules(self) -> tuple[str, ...]:
        return tuple(sorted(self._adblock_rules))

    def is_blocked(self, url: str) -> bool:
        lower_url = url.lower()
        return any(pattern in lower_url for pattern in self._adblock_rules)

    def queue_download(self, download_id: str, url: str) -> DownloadJob:
        if download_id in self._downloads:
            raise ValueError(f'Download with id "{download_id}" already exists.')

        job = DownloadJob(id=download_id, url=url)
        self._downloads[download_id] = job
        return job

    def update_download(self, download_id: str, status: str, saved_path: str | None = None) -> DownloadJob:
        if download_id not in self._downloads:
            raise ValueError(f"Unknown download: {download_id}")

        job = self._downloads[download_id]
        job.status = status
        job.saved_path = saved_path
        return job

    @property
    def downloads(self) -> tuple[DownloadJob, ...]:
        return tuple(self._downloads.values())

    def webview_ipc(self, tab_id: str) -> dict[str, object]:
        if tab_id not in self._tabs:
            raise ValueError(f"Unknown tab: {tab_id}")

        from .ipc import create_webview_ipc

        tab = self._tabs[tab_id]
        return create_webview_ipc(
            tab_id=tab.id, url=tab.url, ad_blocked=self.is_blocked(tab.url)
        ).to_dict()

    def to_dict(self) -> dict[str, object]:
        return {
            "maker": SLATE_MAKER,
            "tabs": [asdict(tab) for tab in self.tabs],
            "bookmarks": [asdict(bookmark) for bookmark in self.bookmarks],
            "history": [asdict(entry) for entry in self.history()],
            "downloads": [asdict(download) for download in self.downloads],
            "ad_block_rules": list(self.adblock_rules),
        }

    @classmethod
    def from_dict(cls, payload: dict[str, object]) -> "BrowserModule":
        browser = cls()

        for item in payload.get("tabs", []):
            tab = BrowserTab(**item)
            browser._tabs[tab.id] = tab

        for item in payload.get("bookmarks", []):
            bookmark = Bookmark(**item)
            browser._bookmarks[bookmark.url] = bookmark

        browser._history = [HistoryEntry(**item) for item in payload.get("history", [])]
        browser._downloads = {
            job.id: job for job in (DownloadJob(**item) for item in payload.get("downloads", []))
        }
        browser._adblock_rules = {str(rule).lower() for rule in payload.get("ad_block_rules", [])}
        return browser
